test/python/pulse/test_parameters.py

```diff
@@ -0,0 +1,159 @@
+        schedule += pulse.SetFrequency(self.alpha, DriveChannel(0))
+        schedule += pulse.ShiftFrequency(self.gamma, DriveChannel(0))
+        schedule += pulse.SetPhase(self.phi, DriveChannel(1))
+        schedule += pulse.ShiftPhase(self.theta, DriveChannel(1))
+
+        schedule.assign_parameters({self.alpha: self.freq, self.gamma: self.shift,
+                                    self.phi: self.phase, self.theta: -self.phase})
+
+        insts = assemble(schedule, self.backend).experiments[0].instructions
+        GHz = 1e9  # pylint: disable=invalid-name
+        self.assertEqual(float(insts[0].frequency*GHz), self.freq)
+        self.assertEqual(float(insts[1].frequency*GHz), self.shift)
+        self.assertEqual(float(insts[2].phase), self.phase)
+        self.assertEqual(float(insts[3].phase), -self.phase)
+
+    def test_multiple_parameters(self):
+        """Expressions of parameters with partial assignment."""
+        schedule = pulse.Schedule()
+        schedule += pulse.SetFrequency(self.alpha + self.beta, DriveChannel(0))
+        schedule += pulse.ShiftFrequency(self.gamma + self.beta, DriveChannel(0))
+        schedule += pulse.SetPhase(self.phi, DriveChannel(1))
+
+        # Partial bind
+        delta = 1e9
+        schedule.assign_parameters({self.alpha: self.freq - delta})
+        schedule.assign_parameters({self.beta: delta})
+        schedule.assign_parameters({self.gamma: self.shift - delta})
+        schedule.assign_parameters({self.phi: self.phase})
+
+        insts = schedule.instructions
+        self.assertEqual(float(insts[0][1].frequency), self.freq)
+        self.assertEqual(float(insts[1][1].frequency), self.shift)
+        self.assertEqual(float(insts[2][1].phase), self.phase)
+
+    def test_with_function(self):
+        """Test ParameterExpressions formed trivially in a function."""
+        def get_frequency(variable):
+            return 2*variable
+
+        def get_shift(variable):
+            return variable - 1
+
+        schedule = pulse.Schedule()
+        schedule += pulse.SetFrequency(get_frequency(self.alpha), DriveChannel(0))
+        schedule += pulse.ShiftFrequency(get_shift(self.gamma), DriveChannel(0))
+
+        schedule.assign_parameters({self.alpha: self.freq / 2, self.gamma: self.shift + 1})
+
+        insts = schedule.instructions
+        self.assertEqual(float(insts[0][1].frequency), self.freq)
+        self.assertEqual(float(insts[1][1].frequency), self.shift)
+
+    def test_substitution(self):
+        """Test Parameter substitution (vs bind)."""
+        schedule = pulse.Schedule()
+        schedule += pulse.SetFrequency(self.alpha, DriveChannel(0))
+
+        schedule.assign_parameters({self.alpha: 2*self.beta})
+        self.assertEqual(schedule.instructions[0][1].frequency, 2*self.beta)
+        schedule.assign_parameters({self.beta: self.freq / 2})
+        self.assertEqual(float(schedule.instructions[0][1].frequency), self.freq)
+
+    def test_channels(self):
```

Comment:
Could we please add a test for what happens when there are two overlapping instructions with the same parameter as the channel to define this behaviour?

Reply:
Instructions have the same parameter as which channel? Do you mean two instructions with parameterized channels? And what would you like the test to be? Verify that assignment of the parameterized channel index on overlapping instructions causes an error?